README.md
# flutter_price_slider

A component that provides price slider just like binance. It could be used for select percentage of price.

## Demo

You can try it on the demo site

https://wingch.github.io/flutter_price_slider/

![demo](https://raw.githubusercontent.com/WingCH/flutter_price_slider/main/demo.gif)

## How to use

```dart 
import 'package:flutter_price_slider/flutter_price_slider.dart';
```

```dart
FlutterPriceSlider(
  width: 200,
  selectedBoxColor: Color(0xFF2ebd85),
  unselectedBoxColor: Color(0xFFf9f9f9),
  selectedTextColor: Color(0xFF000000),
  unselectedTextColor: Color(0XFF7d8896),
  onSelected: (proportion) {
    print("onSelected: $proportion");
  },
),
```

## Features and bugs

Please file feature requests and bugs at the [issue tracker](https://github.com/WingCH/flutter_price_slider/issues).



CHANGELOG.md
## 1.0.1
initial release

test/flutter_price_slider_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_price_slider/flutter_price_slider.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


flutter_service_worker.js
'use strict';
const MANIFEST = 'flutter-app-manifest';
const TEMP = 'flutter-temp-cache';
const CACHE_NAME = 'flutter-app-cache';
const RESOURCES = {
  "version.json": "4b6db237b3514a88107a422469adfb0f",
"index.html": "6fba2f5ec0990f6dd1181bb232896d8f",
"/": "6fba2f5ec0990f6dd1181bb232896d8f",
"main.dart.js": "946df4796f3c3521812ab7f5e89061b3",
"favicon.png": "5dcef449791fa27946b3d35ad8803796",
"icons/Icon-192.png": "ac9a721a12bbc803b44f645561ecb1e1",
"icons/Icon-512.png": "96e752610906ba2a93c65f8abe1645f1",
"manifest.json": "15f73b7e8a8209c2206210b3ac8dea1b",
".git/config": "4490a148183d761b0462390691b9acbe",
".git/objects/03/eaddffb9c0e55fb7b5f9b378d9134d8d75dd37": "87850ce0a3dd72f458581004b58ac0d6",
".git/objects/32/46ad559eeae0370195978eaed83f1053ee13fd": "a043dbc0a0bda96ce2127799ccc27506",
".git/objects/56/cb34b7f8bb9dfda46cb113ab3ed7677518d641": "e428c355ab4ee88e955cdd6288a897d4",
".git/objects/ac/44603374f0b1228527d3dae87e6d27e708251f": "b59c61fefbba8d249a49c74fb2d17a9b",
".git/objects/da/3fdd197e7aed7e3db13b201ca3c5502af3c700": "c7a5280b633288acab28250539d8f95c",
".git/objects/ab/0e98497a51ead7821d1da35a24968ff314e50f": "557c35fe3928eb2af403d1b3926bb9ba",
".git/objects/e5/951dfb943474a56e611d9923405cd06c2dd28d": "c6fa51103d8db5478e1a43a661f6c68d",
".git/objects/20/5bb5db271c6d8de8399864c7bb9b917f638893": "c993b22f115d7f3ae6d5b7b212806539",
".git/objects/8a/aa46ac1ae21512746f852a42ba87e4165dfdd1": "1d8820d345e38b30de033aa4b5a23e7b",
".git/objects/21/c9f052c1363db308990bc5af6f5fe8172b65fb": "20ac946899309305b9b0baa1271834fb",
".git/objects/72/0e51f36ab80e2c2f6c53ce0c1477a144eddb9c": "eae03d8cdae1c54410f5c4b9b31ae2fe",
".git/objects/88/cfd48dff1169879ba46840804b412fe02fefd6": "e42aaae6a4cbfbc9f6326f1fa9e3380c",
".git/objects/08/5ebb170611ebf60c4f04109483c41555de463f": "b2e066cbe32f03e28fa30584ddd5d6d6",
".git/objects/52/c9348e53d64333a635aa3348f55ee976123076": "95fb8e53f87c3ebb0b994a014302253b",
".git/objects/dd/21625a7a0627486d0ed1a64eb98e7072a79d62": "dea76e555359d3f16eaed0ca9d7a5ca5",
".git/objects/b7/49bfef07473333cf1dd31e9eed89862a5d52aa": "36b4020dca303986cad10924774fb5dc",
".git/objects/a8/beffd3ad4fe54d6cabccf83a05477d6a986cd0": "6677888e4a051c7838b5b240c09f0981",
".git/objects/a1/3837a12450aceaa5c8e807c32e781831d67a8f": "bfe4910ea01eb3d69e9520c3b42a0adf",
".git/objects/e1/19262e3e241918f2ec08b506a1ad2c98a17ac5": "dc29bb91c80e07eb18f77f1a5e259bcf",
".git/objects/46/4ab5882a2234c39b1a4dbad5feba0954478155": "2e52a767dc04391de7b4d0beb32e7fc4",
".git/objects/79/ba7ea0836b93b3f178067bcd0a0945dbc26b3f": "f3e31aec622d6cf63f619aa3a6023103",
".git/objects/8c/012917dab71a26d535ff8c614582477a123a6d": "22939eff7a0cb1fc6bb53c9bf363b949",
".git/HEAD": "5ab7a4355e4c959b0c5c008f202f51ec",
".git/info/exclude": "036208b4a1ab4a235d75c181e685e5a3",
".git/logs/HEAD": "a64ab04b6731103f7af93aef9bdea683",
".git/logs/refs/heads/gh-pages": "609503a0b6c2c2fd9301f1f2973a615a",
".git/logs/refs/remotes/origin/gh-pages": "3546efd144d43309d656e10a3d0527af",
".git/description": "a0a7c3fff21f2aea3cfa1d0316dd816c",
".git/hooks/commit-msg.sample": "579a3c1e12a1e74a98169175fb913012",
".git/hooks/pre-rebase.sample": "56e45f2bcbc8226d2b4200f7c46371bf",
".git/hooks/pre-commit.sample": "305eadbbcd6f6d2567e033ad12aabbc4",
".git/hooks/applypatch-msg.sample": "ce562e08d8098926a3862fc6e7905199",
".git/hooks/fsmonitor-watchman.sample": "ea587b0fae70333bce92257152996e70",
".git/hooks/pre-receive.sample": "2ad18ec82c20af7b5926ed9cea6aeedd",
".git/hooks/prepare-commit-msg.sample": "2b5c047bdb474555e1787db32b2d2fc5",
".git/hooks/post-update.sample": "2b7ea5cee3c49ff53d41e00785eb974c",
".git/hooks/pre-merge-commit.sample": "39cb268e2a85d436b9eb6f47614c3cbc",
".git/hooks/pre-applypatch.sample": "054f9ffb8bfe04a599751cc757226dda",
".git/hooks/pre-push.sample": "2c642152299a94e05ea26eae11993b13",
".git/hooks/update.sample": "647ae13c682f7827c22f5fc08a03674e",
".git/hooks/push-to-checkout.sample": "c7ab00c7784efeadad3ae9b228d4b4db",
".git/refs/heads/gh-pages": "6df10d4ea6bdd21f42b140d69f0cb734",
".git/refs/remotes/origin/gh-pages": "6df10d4ea6bdd21f42b140d69f0cb734",
".git/index": "b45bdb162e2c32c817edd3893bcff578",
".git/COMMIT_EDITMSG": "7081fc220d110a41cfb458a10539d7c8",
"assets/AssetManifest.json": "2efbb41d7877d10aac9d091f58ccd7b9",
"assets/NOTICES": "eee37f0ed23e326e3ca321893c462940",
"assets/FontManifest.json": "dc3d03800ccca4601324923c0b1d6d57",
"assets/packages/cupertino_icons/assets/CupertinoIcons.ttf": "6d342eb68f170c97609e9da345464e5e",
"assets/fonts/MaterialIcons-Regular.otf": "4e6447691c9509f7acdbf8a931a85ca1"
};

// The application shell files that are downloaded before a service worker can
// start.
const CORE = [
  "/",
"main.dart.js",
"index.html",
"assets/NOTICES",
"assets/AssetManifest.json",
"assets/FontManifest.json"];
// During install, the TEMP cache is populated with the application shell files.
self.addEventListener("install", (event) => {
  self.skipWaiting();
  return event.waitUntil(
    caches.open(TEMP).then((cache) => {
      return cache.addAll(
        CORE.map((value) => new Request(value, {'cache': 'reload'})));
    })
  );
});

// During activate, the cache is populated with the temp files downloaded in
// install. If this service worker is upgrading from one with a saved
// MANIFEST, then use this to retain unchanged resource files.
self.addEventListener("activate", function(event) {
  return event.waitUntil(async function() {
    try {
      var contentCache = await caches.open(CACHE_NAME);
      var tempCache = await caches.open(TEMP);
      var manifestCache = await caches.open(MANIFEST);
      var manifest = await manifestCache.match('manifest');
      // When there is no prior manifest, clear the entire cache.
      if (!manifest) {
        await caches.delete(CACHE_NAME);
        contentCache = await caches.open(CACHE_NAME);
        for (var request of await tempCache.keys()) {
          var response = await tempCache.match(request);
          await contentCache.put(request, response);
        }
        await caches.delete(TEMP);
        // Save the manifest to make future upgrades efficient.
        await manifestCache.put('manifest', new Response(JSON.stringify(RESOURCES)));
        return;
      }
      var oldManifest = await manifest.json();
      var origin = self.location.origin;
      for (var request of await contentCache.keys()) {
        var key = request.url.substring(origin.length + 1);
        if (key == "") {
          key = "/";
        }
        // If a resource from the old manifest is not in the new cache, or if
        // the MD5 sum has changed, delete it. Otherwise the resource is left
        // in the cache and can be reused by the new service worker.
        if (!RESOURCES[key] || RESOURCES[key] != oldManifest[key]) {
          await contentCache.delete(request);
        }
      }
      // Populate the cache with the app shell TEMP files, potentially overwriting
      // cache files preserved above.
      for (var request of await tempCache.keys()) {
        var response = await tempCache.match(request);
        await contentCache.put(request, response);
      }
      await caches.delete(TEMP);
      // Save the manifest to make future upgrades efficient.
      await manifestCache.put('manifest', new Response(JSON.stringify(RESOURCES)));
      return;
    } catch (err) {
      // On an unhandled exception the state of the cache cannot be guaranteed.
      console.error('Failed to upgrade service worker: ' + err);
      await caches.delete(CACHE_NAME);
      await caches.delete(TEMP);
      await caches.delete(MANIFEST);
    }
  }());
});

// The fetch handler redirects requests for RESOURCE files to the service
// worker cache.
self.addEventListener("fetch", (event) => {
  if (event.request.method !== 'GET') {
    return;
  }
  var origin = self.location.origin;
  var key = event.request.url.substring(origin.length + 1);
  // Redirect URLs to the index.html
  if (key.indexOf('?v=') != -1) {
    key = key.split('?v=')[0];
  }
  if (event.request.url == origin || event.request.url.startsWith(origin + '/#') || key == '') {
    key = '/';
  }
  // If the URL is not the RESOURCE list then return to signal that the
  // browser should take over.
  if (!RESOURCES[key]) {
    return;
  }
  // If the URL is the index.html, perform an online-first request.
  if (key == '/') {
    return onlineFirst(event);
  }
  event.respondWith(caches.open(CACHE_NAME)
    .then((cache) =>  {
      return cache.match(event.request).then((response) => {
        // Either respond with the cached resource, or perform a fetch and
        // lazily populate the cache.
        return response || fetch(event.request).then((response) => {
          cache.put(event.request, response.clone());
          return response;
        });
      })
    })
  );
});

self.addEventListener('message', (event) => {
  // SkipWaiting can be used to immediately activate a waiting service worker.
  // This will also require a page refresh triggered by the main worker.
  if (event.data === 'skipWaiting') {
    self.skipWaiting();
    return;
  }
  if (event.data === 'downloadOffline') {
    downloadOffline();
    return;
  }
});

// Download offline will check the RESOURCES for all files not in the cache
// and populate them.
async function downloadOffline() {
  var resources = [];
  var contentCache = await caches.open(CACHE_NAME);
  var currentContent = {};
  for (var request of await contentCache.keys()) {
    var key = request.url.substring(origin.length + 1);
    if (key == "") {
      key = "/";
    }
    currentContent[key] = true;
  }
  for (var resourceKey of Object.keys(RESOURCES)) {
    if (!currentContent[resourceKey]) {
      resources.push(resourceKey);
    }
  }
  return contentCache.addAll(resources);
}

// Attempt to download the resource online before falling back to
// the offline cache.
function onlineFirst(event) {
  return event.respondWith(
    fetch(event.request).then((response) => {
      return caches.open(CACHE_NAME).then((cache) => {
        cache.put(event.request, response.clone());
        return response;
      });
    }).catch((error) => {
      return caches.open(CACHE_NAME).then((cache) => {
        return cache.match(event.request).then((response) => {
          if (response != null) {
            return response;
          }
          throw error;
        });
      });
    })
  );
}


lib/flutter_price_slider.dart
library flutter_price_slider;

import 'package:flutter/material.dart';

class FlutterPriceSlider extends StatefulWidget {
  const FlutterPriceSlider({
    Key? key,
    this.width = 200,
    required this.selectedBoxColor,
    required this.unselectedBoxColor,
    required this.selectedTextColor,
    required this.unselectedTextColor,
    required this.onSelected,
  }) : super(key: key);

  final double width;
  final Color selectedBoxColor;
  final Color unselectedBoxColor;
  final Color selectedTextColor;
  final Color unselectedTextColor;
  final Function(double) onSelected;

  @override
  _FlutterPriceSliderState createState() => _FlutterPriceSliderState();
}

class _FlutterPriceSliderState extends State<FlutterPriceSlider> {
  List<double> proportions = [0.25, 0.5, 0.75, 1];
  double selectedProportion = 0;

  void onSelected(double proportion) {
    setState(() {
      selectedProportion = proportion;
    });
    widget.onSelected(selectedProportion);
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      width: widget.width,
      child: LayoutBuilder(
        builder: (_, constraints) {
          double maxWidth = constraints.maxWidth;
          return GestureDetector(
            onTap: () {},
            behavior: HitTestBehavior.deferToChild,
            onPanUpdate: (details) {
              // start point (left top): 1.0
              double dx = details.localPosition.dx;
              if (dx < 0) {
                onSelected(0);
                return;
              }
              // Calculate the ratio relative to the parent component
              // maxWidth = 220, dx = 148.5714285714858
              // relativePosition = 0.6341463414634145
              double relativePosition = dx / maxWidth;
              // Find the nearest [0.25, 0.5, 0.75, 1] of relativePosition
              // relativePosition = 0.6341463414634145
              // closetsProportion = 0.75
              final double closetsProportion = proportions.reduce((a, b) {
                return (a - relativePosition).abs() <
                        (b - relativePosition).abs()
                    ? a
                    : b;
              });
              onSelected(closetsProportion);
            },
            child: Row(
              children: proportions
                  .asMap()
                  .map((index, e) {
                    int selectedIndex = proportions.indexWhere((element) {
                      return element == selectedProportion;
                    });
                    return MapEntry(
                      index,
                      Expanded(
                        child: Row(
                          children: [
                            Expanded(
                              child: ProportionSliderItem(
                                proportion: e,
                                boxColor: selectedIndex == -1
                                    ? widget.unselectedBoxColor
                                    : selectedIndex >= index
                                        ? widget.selectedBoxColor
                                        : widget.unselectedBoxColor,
                                textColor: selectedProportion == e
                                    ? widget.selectedTextColor
                                    : widget.unselectedTextColor,
                                onSelect: () => onSelected(e),
                              ),
                            ),
                            if (index != proportions.length - 1)
                              SizedBox(width: 5)
                          ],
                        ),
                      ),
                    );
                  })
                  .values
                  .toList(),
            ),
          );
        },
      ),
    );
  }
}

class ProportionSliderItem extends StatelessWidget {
  const ProportionSliderItem({
    Key? key,
    required this.proportion,
    required this.boxColor,
    required this.textColor,
    required this.onSelect,
  }) : super(key: key);

  final double proportion;
  final Color boxColor;
  final Color textColor;
  final Function() onSelect;

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTap: onSelect,
      behavior: HitTestBehavior.opaque,
      child: Column(
        mainAxisSize: MainAxisSize.min,
        children: [
          Container(
            height: 10,
            decoration: BoxDecoration(
              color: boxColor,
              borderRadius: BorderRadius.all(Radius.circular(1)),
            ),
          ),
          SizedBox(height: 5),
          Text(
            '${proportion * 100}%',
            style: TextStyle(color: textColor),
          )
        ],
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_price_slider/flutter_price_slider.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        appBar: AppBar(
          title: Text("Flutter price slider"),
        ),
        body: Column(
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: [
            Expanded(
              child: Container(
                color: Color(0xFFfffffe),
                child: Column(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: [
                    Text(
                      "Light mode",
                      style: Theme.of(context).textTheme.headline4,
                    ),
                    SizedBox(height: 10),
                    FlutterPriceSlider(
                      width: 200,
                      selectedBoxColor: Color(0xFF2ebd85),
                      unselectedBoxColor: Color(0xFFf9f9f9),
                      selectedTextColor: Color(0xFF000000),
                      unselectedTextColor: Color(0XFF7d8896),
                      onSelected: (proportion) {
                        print("onSelected: $proportion");
                      },
                    ),
                  ],
                ),
              ),
            ),
            Expanded(
              child: Container(
                color: Color(0xFF20262f),
                child: Column(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: [
                    Text(
                      "Dark mode",
                      style: Theme.of(context).textTheme.headline4?.copyWith(
                            color: Colors.white70,
                          ),
                    ),
                    SizedBox(height: 10),
                    FlutterPriceSlider(
                      width: 200,
                      selectedBoxColor: Color(0xFF2dbd85),
                      unselectedBoxColor: Color(0xFF29303d),
                      selectedTextColor: Color(0xFFf1f4f6),
                      unselectedTextColor: Color(0XFF7f8997),
                      onSelected: (proportion) {
                        print("onSelected: $proportion");
                      },
                    ),
                  ],
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}
